C++: IR field flow

MathiasVP opened the pull request:
This PR takes the the first stab at implementing field flow in the IR.

This is still WIP, but any comments is appreciated. One thing I remember is @jbj mentioning that `DefinitionByReferenceNode` couldn't be made a `PostUpdateNode` because there's not a corresponding pre update node to pick, but it seemed like this was the most obvious way to model partial flow to the `this` parameter in setters. And for now https://github.com/Semmle/ql/pull/2921 only reports a couple of problems with nodes missing locations, which shouldn't be a problem to fix.

The failing testcases are from the `Semmle/code`, where we flag a new bad case for `cpp/user-controlled-bypass`. When we are happy with the state of this PR I will make an internal PR and update that test.

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -204,6 +216,27 @@ abstract class PostUpdateNode extends InstructionNode {
+abstract private class PartialDefinitionNode extends PostUpdateNode, TInstructionNode {
+  final Instruction getInstructionOrChi() {
+    exists(ChiInstruction chi |
+      // TODO: This should be a non-conflated ChiInstruction once #3123 is merged
```
It's now merged

jbj replied:
Not it's not

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -321,6 +354,13 @@ predicate localFlowStep(Node nodeFrom, Node nodeTo) { simpleLocalFlowStep(nodeFr
   simpleInstructionLocalFlowStep(nodeFrom.asInstruction(), nodeTo.asInstruction())
+  or
+  exists(LoadInstruction load |
+    // TODO: These can probably be getSourceValue() after #3112 is merged
```
It's now merged (that's the line I meant to comment on a minute ago).

MathiasVP replied:
Sadly my comment turned out to be incorrect. The flow in the following program is not captured by only following exact `Chi` operands:

```c
void sink(int *o);

struct B
{
  int *c;
  void set(int *c) { this->c = c; }
};

void f7(B *b)
{
  b->set(new int);
}

void f8()
{
  B *b = new B();
  f7(b);
  sink(b->c); // flow
}
```

since the load on `b->c` only is a total overlap.

jbj replied:
Then you'll want to merge #3097 and use `isResultConflated` here.

MathiasVP commented:
Here's a quick rundown of the changes in the output:

- `AuthenticationBypass.qlref` in `Semmle/code` has 1 new true positive
- `ql/cpp/ql/test/library-tests/dataflow/dataflow-tests/test_ir.ql` has 3 new true positives and 2 new false positives
- The changes to `ql/cpp/ql/test/library-tests/dataflow/taint-tests/test_diff.ql` and `ql/cpp/ql/test/library-tests/dataflow/taint-tests/test_ir.ql` looks like the expected output given that it's tracking taint.

jbj commented:
> `ql/cpp/ql/test/library-tests/dataflow/dataflow-tests/test_ir.ql` has 3 new true positives and 2 new false positives (the 2 false positives were also present in the old field flow library according to the comments)

And those two false positives are caused by limitations of the shared field-flow library rather than the C++ instantiation of it. They'll be present in the other languages too.

MathiasVP commented:
Testing locally on `Wireshark/Wireshark` revealed something that looks like a bad join order in `DataFlowImpl::parameterThroughFlowCand`:

```
[2020-03-27 10:13:23] (3432s) Tuple counts for DataFlowImpl::parameterThroughFlowCand#ff#antijoin_rhs:
                      2981323763 ~0%     {3} r1 = JOIN DataFlowImpl::parameterThroughFlowCand#ff#shared AS L WITH DataFlowUtil::InstructionNode::getEnclosingCallable_dispred#ff_10#join_rhs AS R ON FIRST 1 OUTPUT L.<2>, L.<1>, R.<1>
                      2968445848 ~2%     {4} r2 = JOIN r1 WITH DataFlowImplCommon::Cached::TParamUpdate#2#ff_10#join_rhs AS R ON FIRST 1 OUTPUT r1.<2>, R.<1>, r1.<1>, r1.<0>
                      832264     ~0%     {3} r3 = JOIN r2 WITH project#DataFlowUtil::ParameterNode::isParameterOf_dispred#2#fff AS R ON FIRST 2 OUTPUT r2.<2>, r2.<3>, r2.<0>
                                         return r3
```

I'm looking into this now.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -219,7 +219,7 @@ abstract class PostUpdateNode extends InstructionNode {
   final Instruction getInstructionOrChi() {
     exists(ChiInstruction chi |
-      // TODO: This should be a non-conflated ChiInstruction once #3123 is merged
+      not chi.isResultConflated() and
```
With this change we lose flow in the following case:

```cpp
void sink(int *);

struct B
{
  int *i;
};

void f7(B* b)
{
  b->i = source();
}
void f8()
{
  B* b = new B;
  f7(b);
  sink(b->i); // flow [not detected]
}
```

Because the `BufferMayWriteSideEffect` for `b` in `f7(b)` is melded into all aliased memory:

```
#   15|     r15_1(glval<unknown>) = FunctionAddress[f7]           :
#   15|     r15_2(glval<B *>)     = VariableAddress[b]            :
#   15|     r15_3(B *)            = Load                          : &:r15_2, m14_8
#   15|     v15_4(void)           = Call                          : func:r15_1, 0:r15_3
#   15|     m15_5(unknown)        = ^CallSideEffect               : ~m14_6
#   15|     m15_6(unknown)        = Chi                           : total:m14_6, partial:m15_5
#   15|     v15_7(void)           = ^BufferReadSideEffect[0]      : &:r15_3, ~m15_6
#   15|     m15_8(unknown)        = ^BufferMayWriteSideEffect[0]  : &:r15_3
#   15|     m15_9(unknown)        = Chi                           : total:m15_6, partial:m15_8
#   16|     r16_1(glval<unknown>) = FunctionAddress[sink]         :
#   16|     r16_2(glval<B *>)     = VariableAddress[b]            :
#   16|     r16_3(B *)            = Load                          : &:r16_2, m14_8
#   16|     r16_4(glval<int *>)   = FieldAddress[i]               : r16_3
#   16|     r16_5(int *)          = Load                          : &:r16_4, ~m15_9
#   16|     v16_6(void)           = Call                          : func:r16_1, 0:r16_5
#   16|     m16_7(unknown)        = ^CallSideEffect               : ~m15_9
```

I'm not totally sure how to recover this flow without accepting flow through `Chi` instructions that update all aliased memory.

Interestingly, the flow is reported in the following modified program:

```cpp
void sink(int *);

struct B
{
  int *i;
};

void f7(B* b)
{
  b->i = source();
}
void f8(B* b)
{
  f7(b);
  sink(b->i); // flow
}
```

since the `Chi` following the `BufferMayWriteSideEffect` does not update all aliased memory:

```
#   15|     r15_1(glval<unknown>) = FunctionAddress[f7]          :
#   15|     r15_2(glval<B *>)     = VariableAddress[b]           :
#   15|     r15_3(B *)            = Load                         : &:r15_2, m12_7
#   15|     v15_4(void)           = Call                         : func:r15_1, 0:r15_3
#   15|     m15_5(unknown)        = ^CallSideEffect              : ~m12_4
#   15|     m15_6(unknown)        = Chi                          : total:m12_4, partial:m15_5
#   15|     v15_7(void)           = ^BufferReadSideEffect[0]     : &:r15_3, ~m12_9
#   15|     m15_8(unknown)        = ^BufferMayWriteSideEffect[0] : &:r15_3
#   15|     m15_9(int *)          = Chi                          : total:m12_9, partial:m15_8
#   16|     r16_1(glval<unknown>) = FunctionAddress[sink]        :
#   16|     r16_2(glval<B *>)     = VariableAddress[b]           :
#   16|     r16_3(B *)            = Load                         : &:r16_2, m12_7
#   16|     r16_4(glval<int *>)   = FieldAddress[i]              : r16_3
#   16|     r16_5(int *)          = Load                         : &:r16_4, ~m15_9
#   16|     v16_6(void)           = Call                         : func:r16_1, 0:r16_5
#   16|     m16_7(unknown)        = ^CallSideEffect              : ~m15_6
```

jbj replied:
Is that a flaw in our support for allocations (https://github.com/Semmle/ql/pull/2797)? I would think that `new B` and a `B *` parameter should be treated the same by the current (unsound) alias analysis.

MathiasVP replied:
Indeed, a fresh allocation certainly shouldn't alias more stuff than a `B*` parameter. I'll look at this issue later today.

MathiasVP replied:
`git bisect`'ing revealed that 9cc3cda58eac4b4b4ad1fc7729a19ed182327faf actually was the commit that removed the allocation from all aliased memory.

MathiasVP replied:
Merging in https://github.com/Semmle/ql/pull/3171 fixes the problem.

MathiasVP commented:
CPP-difference run: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/996

jbj commented:
> CPP-difference run: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/996

Most of the differences on openjdk come from #3123. The two changes for `cpp/signed-overflow-check` are also from a different PR, but I forget which.

MathiasVP commented:
> Most of the differences on openjdk come from #3123. The two changes for `cpp/signed-overflow-check` are also from a different PR, but I forget which.

Oh yeah, good point. I should have chosen my baseline commit more carefully. I'll create a new one that compares against master

jbj commented:
I don't think you want to compare against master since some of the unrelated changes come from unmerged PRs. You want the baseline to be a branch with all the PRs that you've included in this PR (on top of the merge-base to master).

MathiasVP commented:
> I don't think you want to compare against master since some of the unrelated changes come from unmerged PRs. You want the baseline to be a branch with all the PRs that you've included in this PR (on top of the merge-base to master).

I don't think I have included any unmerged PRs in here yet? So far I've only merged from master. So #3123 isn't actually included in this PR. Only the part in #3097.

MathiasVP commented:
Here's a more honest CPP-differences run: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/999

MathiasVP commented:
I've pushed a new commit that modifies the `PostUpdateNodes` so that they're attached to `Chi` instructions now. It seems like this loses some flows, but merging in #2704 seems to fix this.

MathiasVP commented on cpp/ql/test/library-tests/dataflow/dataflow-tests/dataflow-ir-consistency.expected:
```diff
@@ -30,6 +30,9 @@ localCallNodes
 postIsNotPre
 postHasUniquePre
 uniquePostUpdate
+| ref.cpp:83:5:83:17 | Chi | Node has multiple PostUpdateNodes. |
+| ref.cpp:100:34:100:36 | InitializeIndirection | Node has multiple PostUpdateNodes. |
+| ref.cpp:109:5:109:22 | Chi | Node has multiple PostUpdateNodes. |
```
I've described the source of these consistency failures [here](https://github.slack.com/archives/CP0LHP150/p1586173820201600). According to aschackmull it's not clear whether this actually causes any problems.

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -321,6 +385,13 @@ predicate localFlowStep(Node nodeFrom, Node nodeTo) { simpleLocalFlowStep(nodeFr
+    nodeTo.asInstruction() = load and
+    nodeFrom.asInstruction() = chi and
+    load.getSourceValueOperand().getAnyDef() = chi
+  )
```
You can move this case into `simpleInstructionLocalFlowStep`. That will save you two `asInstruction` calls, and it will make this code `cached` when #3189 is merged.

MathiasVP replied:
I've done this now in d56284f. This case I've added in `simpleInstructionLocalFlowStep` has an obvious generalization, which we might add at some later stage. But I think that deserves its own investigation first.

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -346,6 +417,11 @@ private predicate simpleInstructionLocalFlowStep(Instruction iFrom, Instruction
+  // TODO: To capture flow from a partial definition of an object (i.e., a field write) to the object
+  // we add dataflow through partial chi operands, but only if the chi node is not the chi node for all
+  // aliased memory.
+  iTo.getAnOperand().(ChiPartialOperand).getDef() = iFrom and not iTo.isResultConflated()
```
I worry that the new cases will lead to field conflation like I saw in https://github.com/Semmle/ql/pull/3123#issuecomment-609818877.

Given `struct S { int x, y; }`, this case adds flow from a write to `s.x` into `s` itself. The new case for chi -> load above will in principle add flow from `s` to a read of `s.y`. In practice, we're likely to be saved because the aliased SSA will usually not model `s.y` as a partial read from `s`. But I'm sure it will sometimes. What if there's a copy `S s2 = s;` followed by `s2.x = unrelated` to create a chi node and then `sink(s2.y)`? Or what if `s` is sent across a function call?

MathiasVP replied:
I've actually noticed that since we moved to PostUpdateNodes on `ChiInstructions` we can actually have a _tiny_ amount of field flow without this addition. So I've removed this case in d56284f in the hopes that we can get this into 1.24.

jbj replied:
If what you need is flow out of definition-by-reference, then you can perhaps do for data flow what I did for taint in https://github.com/Semmle/ql/pull/3219. That PR splits the `ChiPartialOperand` case in two, and the first of those two cases might be suitable for data flow rather than just taint.

MathiasVP replied:
Won't that have the same problems regarding field conflation in the following program?
```cpp
void f(S* s) {
  s->x = source();
}

void g(S* s) {
  f(s);
  // s = chi(total: Uninitialized[s], partial: source())
  sink(s->y)
}
```

jbj replied:
Does this conflation require flow through output params?

I'd hope that with the first rule of #3219 (not the second) the `s->x = source()` assignment does not cause flow into `*s` (the whole struct). There is a phi node after that assignment, right?

MathiasVP replied:
You're right. It doesn't cause flow into the whole struct. The flow through output params was just to generate a `WriteSideEffectInstruction`.

It does seem like adding the first flow rule from #3219 gives us a few additional results in various library tests. I'll create a separate PR for that.

MathiasVP replied:
I've opened https://github.com/Semmle/ql/pull/3220. AFAIC it's not related to this PR, but the new flows in `library-tests` looks like we want to have this anyway.

MathiasVP commented:
A fresh CPP-differences run: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1020/

jbj commented:
> A fresh CPP-differences run: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1020/

It's not all that fresh -- it seems to be based on an old `master` from 2 April. There's clearly some performance work to be done. Be sure to merge from the latest `master`, including path explanations, before investigating performance.

MathiasVP commented:
> It's not all that fresh -- it seems to be based on an old `master` from 2 April. There's clearly some performance work to be done. Be sure to merge from the latest `master`, including path explanations, before investigating performance.

Will do. Fresh was the wrong choice of word here. I meant fresh as in "a new one based on the changes made to this PR".

MathiasVP commented:
The latest [CPP-Differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1030/) revealed a performance regression due to the following predicate in `DataFlowImplCommon.qll`:

```codeql
pragma[nomagic]
private predicate parameterValueFlow0(
  ParameterNode p, Node node, ContentOption contentIn, ContentOption contentOut
) {
  p = node and
  Cand::cand(p, _) and
  contentIn = TContentNone() and
  contentOut = TContentNone()
  or
  // local flow
  exists(Node mid |
    parameterValueFlow(p, mid, contentIn, contentOut) and
    LocalFlowBigStep::localFlowBigStep(mid, node)
  )
  or
  ...
}
```

with this PR we get the following tuple counts for iteration 5:

```
(2455s) Tuple counts for DataFlowImplCommon::Cached::FlowThrough::Final::parameterValueFlow0#ffff#cur_delta:
1579748   ~0%         {4} r1 = SCAN DataFlowImplCommon::Cached::FlowThrough::Final::parameterValueFlow#ffff#prev_delta AS I OUTPUT I.<1>, I.<0>, I.<2>, I.<3>
109572057 ~11994%     {4} r2 = JOIN r1 WITH DataFlowImplCommon::Cached::FlowThrough::LocalFlowBigStep::localFlowBigStep#ff AS R ON FIRST 1 OUTPUT R.<1>, r1.<1>, r1.<2>, r1.<3>
...
109577945 ~11984%     {4} r15 = r2 \/ r14
...
109580255 ~11974%     {4} r27 = r15 \/ r26
...
109580255 ~11974%     {4} r30 = r27 \/ r29
...
109580255 ~11974%     {4} r33 = r30 \/ r32
...
109580255 ~11974%     {4} r36 = r33 \/ r35
...
109580255 ~11974%     {4} r39 = r36 \/ r38
...
109580255 ~11974%     {4} r42 = r39 \/ r41
...
109580255 ~11974%     {4} r45 = r42 \/ r44
14493     ~4%         {4} r46 = r45 AND NOT DataFlowImplCommon::Cached::FlowThrough::Final::parameterValueFlow0#ffff#prev AS R(r45.<1>, r45.<0>, r45.<2>, r45.<3>)
14493     ~6%         {4} r47 = SCAN r46 OUTPUT r46.<1>, r46.<0>, r46.<2>, r46.<3>
                    return r47
```

where's on the current `master` (6e88b6d4f462523cd7313e870475124f428e84e0) we get:

```
(61s) Tuple counts for DataFlowImplCommon::Cached::FlowThrough::Final::parameterValueFlow0#ffff#cur_delta:
57061  ~0%     {4} r1 = SELECT DataFlowImplCommon::Cached::FlowThrough::Final::parameterValueFlow#ffff#prev_delta AS I ON I.<2> = I.<3>
57061  ~3%     {3} r2 = SCAN r1 OUTPUT r1.<1>, r1.<0>, r1.<2>
0      ~0%     {3} r3 = JOIN r2 WITH DataFlowImplCommon::Cached::FlowThrough::LocalFlowBigStep::localFlowBigStep#ff AS R ON FIRST 1 OUTPUT r2.<1>, r2.<2>, R.<1>
...
0      ~0%     {3} r9 = r3 \/ r8
...
0      ~0%     {3} r15 = r9 \/ r14
...
0      ~0%     {3} r20 = r15 \/ r19
...
0      ~0%     {3} r25 = r20 \/ r24
...
0      ~0%     {3} r28 = r25 \/ r27
...
0      ~0%     {3} r31 = r28 \/ r30
106958 ~0%     {4} r32 = SELECT DataFlowImplCommon::Cached::FlowThrough::Final::parameterValueFlow0#ffff#prev AS I ON I.<2> = I.<3>
106958 ~5%     {3} r33 = SCAN r32 OUTPUT r32.<0>, r32.<1>, r32.<2>
106958         {3} r34 = MATERIALIZE r33 AS antijoin_rhs
0      ~0%     {3} r35 = r31 AND NOT r34(r31.<0>, r31.<2>, r31.<1>)
0      ~0%     {4} r36 = SCAN r35 OUTPUT r35.<0>, r35.<2>, r35.<1>, r35.<1>
                return r36
```

I'm not exactly sure what's causing the change in RA. On `master` we only keep the rows where `contentIn` and `contentOut` are equal, where's this is not the case after the changes in this PR.

Here's the relevant predicate differences:

Master:
```codeql
  exists(/* DataFlowUtil::Node */ TIRDataFlowNode mid |
    exists(/* DataFlowImplCommon::ContentOption */ TContentOption arg3 |
      rec DataFlowImplCommon::Cached::FlowThrough::Final::parameterValueFlow#ffff(p,
                                                                                  mid,
                                                                                  contentOut,
                                                                                  arg3),
      contentOut = arg3
    ),
    DataFlowImplCommon::Cached::FlowThrough::LocalFlowBigStep::localFlowBigStep#ff(mid,
                                                                                   node)
  );
```

This PR:
```codeql
exists(/* DataFlowUtil::Node */ TIRDataFlowNode mid |
  DataFlowImplCommon::Cached::FlowThrough::LocalFlowBigStep::localFlowBigStep#ff(mid,
                                                                                 node),
  rec DataFlowImplCommon::Cached::FlowThrough::Final::parameterValueFlow#ffff(p,
                                                                              mid,
                                                                              contentIn,
                                                                              contentOut)
);
```

rdmarsh2 commented:
> I'm not exactly sure what's causing the change in RA. On master we only keep the rows where `contentIn` and `contentOut` are equal, where's this is not the case after the changes in this PR.

I think the constraint on master is inferred indirectly from `readStep` and `storeStep` being `none()`, and which isn't true after the PR. The bigger problem is that there are many possible `mid`  nodes that connect a pair of `p` and `node` with the same content. That might be coming from chi nodes on parameters, but averaging 100 of those seems extreme. Could it be caused by alias/conflation of parameters?

MathiasVP commented:
> I think the constraint on master is inferred indirectly from `readStep` and `storeStep` being `none()`, which isn't true after the PR. The real problem is that there are many possible `mid` nodes that connect a pair of `p` and `node` with the same content. That might be coming from chi nodes on parameters, but averaging 100 of those seems extreme. Could it be caused by alias/conflation of parameters?

The root cause was a was a stupid mistake on my side. In a previous commit I introduced an abstract class for `ParameterNode`s (to split explicit and implicit parameters), and then removed the abstract'ness from the class following your comment in https://github.com/Semmle/ql/pull/3123. But (unlike Jonas) I forgot to introduce a char pred to the `ParameterNode` class, so `parameterValueFlow0` was not restricted to only parameters.

I've fixed this and started a new CPP-differences.

MathiasVP commented:
Here's the newest CPP-differences after merging in master after Robert's outparam PR (along with necessary flow from #3220): https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1043/

Performance still looks good. I'll take a look at the profile for `microsoft/Chakra` just to be sure. We have a couple of new results now as well:

```
=== Changes per project (+new -fixed) ===

g/git/git                                                                   +1 -0
g/php/php-src                                                               +2 -0

=== Changes per query (+new -fixed) ===

cpp/path-injection                                                          +2 -0

cpp/uncontrolled-allocation-size                                            +1 -0
```

I'll take a look at these tomorrow.

jbj commented:
> Here's the newest CPP-differences after merging in master after Robert's outparam PR (along with necessary flow from #3220): https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1043/

That looks great!

> Performance still looks good. The only major spike in time spend is on `microsoft/Chakra`, but that is due to increased build time.

It looks to me like there's a performance regression beyond the noise level. Looking at the [analysis time diffs](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1043//artifact/analysis_time.html), the first `DefaultTaintTracking`-based queries in the suite are clearly taking longer than they used to. The biggest relative slowdown is on git/git, but that could be noise.

I do expect a slowdown since we're doing more work than before, and a few percent seems reasonable. But do you see any individual predicates that have slowed down and are now slower than they ought to be?

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
````diff
@@ -204,6 +238,38 @@ abstract class PostUpdateNode extends InstructionNode {
+ * setY(&x); // a partial definition of the object `x`.
+ * ```
+ */
+abstract class PartialDefinitionNode extends PostUpdateNode, TInstructionNode { }
````
Can this class be private? If not, can it be non-abstract?

MathiasVP replied:
It's marked as private in the AST-based `DataFlowUtil.qll`, so yes. I've fixed this in 62e2ffe62346a01a4bf63eca9d0b747919f7738f.

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -204,6 +238,38 @@ abstract class PostUpdateNode extends InstructionNode {
+  }
+
+  // There might be multiple `ChiInstructions` that has a particular instruction as
+  // the total operand - so this definition give consistency errors in
```
```suggestion
  // the total operand - so this definition gives consistency errors in
```

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -360,6 +430,14 @@ private predicate simpleInstructionLocalFlowStep(Instruction iFrom, Instruction
+  exists(ChiInstruction chi | iFrom = chi |
+    not chi.isResultConflated() and
+    iTo.(LoadInstruction).getSourceValueOperand().getAnyDef() = chi
+  )
```
Please add comments above these two new cases. Recent experience has shown us that cases like these can interact in unexpected ways as we add more, and good comments will ensure we don't accidentally break them if/when we need to adjust them.

MathiasVP replied:
Good point. I've added this now.

jbj replied:
That's very helpful. Reading those comments makes me think about how we can improve these rules in follow-up PRs, like allowing for phi nodes between the chi and the load. With these examples in the comments, it should be much harder to accidentally break the desired functionality.

jbj commented on cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:
```diff
@@ -85,14 +85,14 @@ void class_field_test() {
+	sink(mc2.a); // [FALSE POSITIVE]
+	sink(mc2.b); // tainted
+	sink(mc2.c); // tainted
+	sink(mc2.d); // [FALSE POSITIVE]
```
I no longer see the `.expected` changes that correspond to these comment changes.

MathiasVP replied:
You're right. I think these results were lost when we limited `PostUpdateNodes` to be on chi instructions with non-conflated results. I've fixed the annotations in 62e2ffe62346a01a4bf63eca9d0b747919f7738f.

MathiasVP commented:
> I do expect a slowdown since we're doing more work than before, and a few percent seems reasonable. But do you see any individual predicates that have slowed down and are now slower than they ought to be?

The query that has the biggest slowdown `cpp/cpp/path-injection` has has the biggest slowdown in the query phase of the evaluation. Looking at the timing stats for that particular stage locally for `git/git` and `wireshark/wireshark` (and a brief look at the timing stats for other stages) there's no predicate that immediately caught my eye as unreasonably slow.

MathiasVP commented:
Here's the relevant internal PR: https://git.semmle.com/Semmle/code/pull/36713
